src/hardware/driver/relay.rs

```diff
@@ -0,0 +1,256 @@
+    mutex: &'a spin::Mutex<Mcp230xx<I2C, Mcp23008>>,
+    delay: asm_delay::AsmDelay,
+    k1_en_n: RelayPin,
+    k1_en: RelayPin,
+    k0_d: RelayPin,
+    k0_cp: RelayPin,
+}
+
+impl<'a, I2C, E> Relay<'a, I2C>
+where
+    I2C: WriteRead<Error = E> + Write<Error = E>,
+    E: Debug,
+{
+    const K0_DELAY: fugit::MillisDuration<u64> =
+        fugit::MillisDurationU64::millis(10);
+    const K1_DELAY: fugit::MillisDuration<u64> =
+        fugit::MillisDurationU64::millis(10);
+
+    /// Construct a new [Relay].
+    ///
+    /// # Args
+    /// * `mutex`   - mutex of a shared MCP23008
+    /// * `ccdr`    - core clocks to construct a `delay`
+    /// * `channel` - Driver channel to construct the [Relay] for
+    pub fn new(
+        mutex: &'a spin::Mutex<Mcp230xx<I2C, Mcp23008>>,
+        ccdr: &rcc::CoreClocks,
+        channel: Channel,
+    ) -> Self {
+        let (k1_en_n, k1_en, k0_d, k0_cp) = if channel == Channel::LowNoise {
+            (
+                RelayPin::LN_K1_EN_N,
+                RelayPin::LN_K1_EN,
+                RelayPin::LN_K0_D,
+                RelayPin::LN_K0_CP,
+            )
+        } else {
+            (
+                RelayPin::HP_K1_EN_N,
+                RelayPin::HP_K1_EN,
+                RelayPin::HP_K0_D,
+                RelayPin::HP_K0_CP,
+            )
+        };
+        let mut mcp = get_mcp(mutex);
+        // set GPIOs to default position
+        mcp.set_gpio(k1_en.into(), Level::Low).unwrap();
+        mcp.set_gpio(k1_en_n.into(), Level::High).unwrap();
+        mcp.set_gpio(k0_d.into(), Level::Low).unwrap();
+        mcp.set_gpio(k0_cp.into(), Level::Low).unwrap();
+
+        let delay = asm_delay::AsmDelay::new(asm_delay::bitrate::Hertz(
+            ccdr.c_ck().to_Hz(),
+        ));
+        Relay {
+            mutex,
+            delay,
+            k1_en_n,
+            k1_en,
+            k0_d,
+            k0_cp,
+        }
+    }
+}
+
+pub mod sm {
+    use super::*;
+    statemachine! {
+        transitions: {
+            *Disabled + Enable / engage_k0 = EnableWaitK0,
+            EnableWaitK0 + RelayDone / disengage_k1 = EnableWaitK1,
+            EnableWaitK1 + RelayDone = Enabled,
+            Enabled + Disable / engage_k1 = DisableWaitK1,
+            DisableWaitK1 + RelayDone / disengage_k0 = DisableWaitK0,
+            DisableWaitK0 + RelayDone = Disabled
+        }
+    }
+}
+
+impl<'a, I2C, E> sm::StateMachineContext for Relay<'a, I2C>
+where
+    I2C: WriteRead<Error = E> + Write<Error = E>,
+    E: Debug,
+{
+    // set K0 to upper position (note that "upper" and "lower" refer to the schematic)
+    fn engage_k0(&mut self) {
+        let mut mcp = get_mcp(self.mutex);
+        // set flipflop data pin
+        mcp.set_gpio(self.k0_d.into(), Level::High).unwrap();
+        // set flipflop clock input low to prepare rising edge
+        mcp.set_gpio(self.k0_cp.into(), Level::Low).unwrap();
+        self.delay.delay_us(100u32);
```

Comment:
100 µs plus I2C transaction time just to clock a flipflop? I think this can be removed. Alternatively commenty why 100 µs.